modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/PartitionStatesMetricSource.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.ignite.internal.metrics.MetricSetBuilder;
+import org.apache.ignite.internal.metrics.MetricSource;
+import org.apache.ignite.internal.replicator.TablePartitionId;
+import org.apache.ignite.raft.jraft.Node;
+import org.apache.ignite.raft.jraft.RaftGroupService;
+import org.jetbrains.annotations.Nullable;
+
+/** Source of metrics for table partition statuses. */
+class PartitionStatesMetricSource implements MetricSource {
+    private final String metricSourceName;
+
+    private final int tableId;
+
+    private final DisasterRecoveryManager disasterRecoveryManager;
+
+    /** Enablement status. Accessed from different threads under synchronization on this object. */
+    private boolean enabled;
+
+    PartitionStatesMetricSource(
+            CatalogTableDescriptor tableDescriptor,
+            DisasterRecoveryManager disasterRecoveryManager
+    ) {
+        this.tableId = tableDescriptor.id();
+        this.disasterRecoveryManager = disasterRecoveryManager;
+
+        metricSourceName = String.format("partition.states.zone.%s.table.%s", tableDescriptor.zoneId(), tableDescriptor.id());
+    }
+
+    @Override
+    public String name() {
+        return metricSourceName;
+    }
+
+    @Override
+    public synchronized @Nullable MetricSet enable() {
+        if (enabled) {
+            return null;
+        }
+
+        var builder = new MetricSetBuilder(metricSourceName);
+
+        builder.longGauge(
+                "UnavailablePartitionCount",
+                "Count of partitions not yet started.",
+                () -> calculatePartitionCountByLocalState(UNAVAILABLE)
+        );
+
+        builder.longGauge(
+                "HealthyPartitionCount",
+                "Count of living partitions with a healthy state machine.",
+                () -> calculatePartitionCountByLocalState(HEALTHY)
+        );
+
+        builder.longGauge(
+                "InitializingPartitionCount",
+                "Count of partitions that are starting right now.",
+                () -> calculatePartitionCountByLocalState(INITIALIZING)
+        );
+
+        builder.longGauge(
+                "InstallingSnapshotPartitionCount",
+                "Count of partitions that installing Raft snapshots from the leader.",
+                () -> calculatePartitionCountByLocalState(INSTALLING_SNAPSHOT)
+        );
+
+        builder.longGauge(
+                "BrokenPartitionCount",
+                "Count of broken partitions.",
+                () -> calculatePartitionCountByLocalState(BROKEN)
+        );
+
+        return builder.build();
+    }
+
+    @Override
+    public synchronized void disable() {
+        enabled = false;
+    }
+
+    @Override
+    public synchronized boolean enabled() {
+        return enabled;
+    }
+
+    private long calculatePartitionCountByLocalState(LocalPartitionStateEnum state) {
+        long[] count = {0};
+
+        disasterRecoveryManager.raftManager.forEach((raftNodeId, raftGroupService) -> {
```

Comment:
I think that this algorithm will be extremely inefficient. Loading metrics for all tables will be `O(n^2)`, which is pretty bad in general. I don't know if I should ask you to fix it now, but you should keep in mind that ideally we should do better than this. If you have better ideas about improving the approach, please reply here, we can discuss them.

Reply:
Today I thought that it would be great to have some kind of listener that would trigger a callback when the status of the node changes (To do this you will need to update the code to jraft), but then we would need to somehow manage this and keep in memory some amount of data that we don’t really need.

I looked more closely at making some kind of `Loza#raftGroupService` to make a loop through partitions, but we will either need to have 2 maps, or it will be more difficult to create `org.apache.ignite.internal.raft.RaftNodeId`.

On the other hand, we need such a metric only for disaster recovery and we may not worry about its performance for now, since we will go through cycles through the entities in memory, what do you think?

Reply:
> I don’t understand where the O(n^2), can you clearly indicate?

Never thought that I would have to do this, but here we go. Suppose you have `N` tables, each of them has an associated metrics source. For each source you have `5` metrics that show partition states. For each state you scan all `N` tables while looking for the one that you need (let's ignore partitions for simplicity). That's called a linear search and it's `O(N)`. If you multiply it by the number of metrics sources and the number of metrics, you will get `5*N*N` read operations per metrics accumulation, which is `O(N^2)`. This is what I meant.

Reply:
Let's figure it out.
N - the number of tables, which is equal to the number of metrics source.
I suggest adding 5 metrics.

For each metric, O(N) will be performed.
Those. for each metric source of the total will be O(5*N), where it will be N^2 I don’t understand.

Reply:
> what do you think?

I think that eventually it will bite us in the rear. Avoiding listeners, the only good solution would be to ask a `TableManager` for all replicas of a given table, and to have it there already grouped by `tableId`. Then we should be able to access the raft group from the replica, that would be optimal and relatively simple.

Reply:
Yes, but on the other hand, in some future we will switch to partition colocation, and then it will be possible to optimize this, what do you think?

Reply:
> what do you think?

Same problem will be there for large numbers of zones, but I think that this situation would be rare. Originally I said I don't know if I should ask you to fix it now_ for this exact reason by the way

> for each metric source of the total will be O(5*N), where it will be N^2 I don’t understand.

What exactly was confusing in my explanation? I thought it was as details as possible. If you have `O(N)` metrics, and all of them require `O(N)` operations to be computed, then computing all of them would be `O(N*N)`. What else should I add?

Reply:
> What exactly was confusing in my explanation?

I think that it is more correct to estimate the complexity of O(MxN), where N is the number of tables and M is the number of metrics?
O(N*N) is obtained if we load all metrics for all tables at once.
If we load metrics only for a specific one or for the entire table, then this will not happen.

I looked in the `TableManager` and did not find a suitable API for getting the state of the raft node; in `org.apache.ignite.internal.table.TableImpl` we store `org.apache.ignite.internal.raft.service.RaftGroupService`, which is different from the `org.apache.ignite.raft.jraft.RaftGroupService` that is currently in use.

To avoid O(N*N) as I described above, we can make metrics common to all tables (in the future zones) and then there will always be O(M*N). I don’t really like adding `org.apache.ignite.raft.jraft.RaftGroupService` to `org.apache.ignite.internal.table.TableImpl`, we can add `org.apache.ignite.internal.raft.RaftNodeId` to `TableImpl`.

Well, or leave it as is, what do you think?

Reply:
> O(N*N) is obtained if we load all metrics for all tables at once.

Yes, that's precisely what I said. Quote: _"Loading metrics for all tables will be O(n^2)"_

> I looked in the TableManager and did not find a suitable API for getting the state of the raft node

That's right, it doesn't exist right now, but it could if we really needed it.

> Well, or leave it as is, what do you think?

We can leave it as is. My main point was that you should be aware of the complexity of your algorithms, and be sure that there's a good reason and it's safe to have O(N*N) code in some places. Ideally, such a code should be commented with explanation why it's OK. Otherwise readers will have same questions as I do